doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,1323 @@
+A second proposal NEP 43 details proposed changes to the universal
+functions.
+Note that only the implementation of both NEPs will provide the desired full
+functionality.
+
+
+.. note::
+
+    At this time this NEP is in a preliminary state. Both internal and
+    external API may be adapted based on user input or implementation needs.
+    The general design principles and choices, while provisional, should not
+    be expected to change dramatically.
+
+
+Detailed Description
+--------------------
+
+NEP 41 layed out the creation of a class hierarchy for datatypes using the
+new DType classes to provide all necessary implementations.
+This NEP defines the specific choice of API necessary to define new DTypes.
+Here these are suggested as C-API slots, however, conceptually these
+translate identically to Python methods.
+
+Additionally, the NEP proposes to implement the notion of *abstract* DTypes.
+Further, we detail – in part – how the proposed methods (C-API slots)
+enable all necessary use cases.
+
+Each section will begin with a short motivation of the issue or what
+problem is addressed. This is followed by a description of the proposed
+design choice, and then may list alternatives.
+
+
+Nomenclature
+""""""""""""
+
+As a brief note on nomenclature, it should be noted that ``dtype`` normally
+denotes the dtype *instance*, which is the object attached to a numpy array.
+On the other hand the ``DType`` class is the subclass of ``np.dtype``.
+On the C-level we currently use the name ``descriptor`` or ``descr`` in many
+cases interchangeably with the dtype instance. This will be used in proposed
+C-API names for clarification.
+Note that the notion of dtype class is currently represented mainly as
+the ``dtype.num`` and ``dtype.char``.
+Please see the `dtype hierarchy figure <hierarchy_figure>`_ for an
+illustration of this distinction.
+
+There are currently classes in NumPy for numeric types e.g. 
+``np.float64``; however,
+these are not DTypes but the corresponding scalar classes
+(see also NEP 40 and 41 for discussion on why these are largely unrelated to
+the proposed changes).
+
+
+Proposed access to DType class
+""""""""""""""""""""""""""""""
+
+**Motivation:**
+
+Currently we often use e.g. ``np.dtype(np.int64)`` to create the dtype
+instance corresponding to a given scalar type. Adding the DType class
+means that a way to access that class may be required.
+
+**Description:**
+
+To avoid duplication, but also expose the classes conveniently to users
+we propose the addition of::
+
+    np.dtype[np.int64]
+
+as a class getter. This can work both for user and NumPy DTypes,
+although, in many cases a library may choose to provide a more direct
+way to access the specific DType class.
+This method may initially be limited to concrete DTypes. 
+The main reason for this choice is to provide a single
+clear and future-proof way to find the DType class given the
+Python (scalar) class.
```

Comment:
Are we sure we want to introduce a second way to spell this, given that `np.dtype(cls)` is already a well-established mechanism to convert a scalar type to a dtype?

Reply:
No, I am not sure about this, it is more of a thought and I could reduce or even remove this for now.

There is a generalization whether we should after all consider to create `parametric` instances (similar to our "S0").
I had decided against this, but passing around parametric instances has its advantages (since for some functions the input can be a `dtype` instance _or_ a `DType` class).
But on the other hand, means that the neat split between: instances can always be attached arrays, classes cannot. Is gone.

Note that there are some subtleties around this. E.g. `np.dtype(np.float64)` enforces that you do get native byte-order (the default instances), while `np.dtype[np.float64]` (or however you spell it if you want to) will *not* enforce that if the input is already a `>float64`.

So the question is whether `np.dtype(Unit)` should be allowed and return something which cannot be attached to an array...

I guess the main issue is about `issubdtype` checks in the long run.  E.g. how to check whether a dtype is `inexact`, since `isinstance(dtype, np.dtype[np.inexact])` is not very cute (of course the `dtype.type` way/subdtype stays valid.